Add new module managing keycloak user

ndclt opened the pull request:
##### SUMMARY
This module (keycloak_user) allows the use of the Keycloak REST API (http://www.keycloak.org/docs-api/3.4/rest-api/index.html) to administrate Keycloak (http://www.keycloak.org) user.

Keycloak is an Open Source Identity and Access Management system spearheaded by Red Hat. It provides OpenID Connect and SAML authentication/authorization services.

Two prior modules in this namespace have been merged (keycloak_client in #31716 and keycloak_clienttemplate in #33419). This module added some code in module_utils/keycloak.py. It also add a file with a function managing the CRUD states and create a new `KeycloakUser` class used by the crud function.

Three other modules currently have new module PRs pending (keycloak_group in #35637 by @adamgoossens,  keycloak_scope in #35558 by @eikef, keycloak_realms in #)#35844; they extend module_utils/keycloak.py as well, though it's all strictly additions.

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
keycloak_user

ndclt committed:
Add new module managing keycloak user

mattclay commented:
CI failure in unit tests: https://app.shippable.com/github/ansible/ansible/runs/107731/6/tests

ndclt commented:
> CI failure in unit tests: https://app.shippable.com/github/ansible/ansible/runs/107731/6/tests

@mattclay thank you for your message. My last commit should avoid the previous failures.

ndclt commented:
The CI failed because of the test under  T=freebsd/12.0/1. When I look at the errors, it does not seem to be related to my code and I don't know what to do.

Am I wrong?

ndclt committed:
dev: add credentials for user

ndclt committed:
doc: add a value to user_secret in example

ndclt committed:
test: check that credentials value is no_log

ndclt committed:
dev: add 'value' in credential key to set at no_log

dazasa commented:
Any news on this? I

ndclt committed:
doc: change key user_secret to value

jamescassell commented on lib/ansible/modules/identity/keycloak/keycloak_user.py:
```diff
@@ -0,0 +1,441 @@
+        description:
+            - user_id of client to be worked on. This is usually an UUID. This and I(client_username)
+              are mutually exclusive.
+        aliases: [ userId ]
```
Is there precedent for this alias? -- generally ansible like snake_case, and I'm not sure there's much camelCase, especially for module arguments.

ndclt replied:
Yes, in other keycloak modules: [for example](https://github.com/ansible/ansible/blob/8e99f0ecc6df7d2c567e4ad802a52ee9974f9cee/lib/ansible/modules/identity/keycloak/keycloak_client.py#L92).

I don't say it is a good choice but only it exists.

jamescassell commented on lib/ansible/modules/identity/keycloak/keycloak_user.py:
```diff
@@ -0,0 +1,441 @@
+    credentials:
+        description:
+            - a dictionary setting the user password.
+        type: dict
```
generally ansible doesn't want to be a thin wrapper around the backing API, so maybe break this into smaller options

ndclt replied:
Ok. It will come in the next days.

ndclt replied:
@jamescassell: this is done in the last commits. Days became months... :disappointed:

jamescassell commented on lib/ansible/modules/identity/keycloak/keycloak_user.py:
```diff
@@ -0,0 +1,441 @@
+        type: str
+
+    required_actions:
+        description:
```
This could use more explanation.

ndclt replied:
Updated in a9e9ecf451b83cfa772e951608d02a5c3acb8feb.

ndclt committed:
update documentation of required_actions

ndclt committed:
add the __metaclass__ in test file

ndclt committed:
dev: use the get_token function

ndclt committed:
doc: change targeted version

ndclt committed:
test: change mocks

ndclt committed:
feature: add only the password as string not dict

ndclt committed:
refactor: use the crud and KeycloakUser class

delete all users reference in the keycloak.py file
create a crud file with a class instance as argument which must have some attributes and function in order to work.

ndclt committed:
test: add new test for setting a new password

ndclt committed:
fix: correctly set the password for a user

ndclt committed:
fix: correctly put the username in the payload

ndclt committed:
fix: compatibility python 2.6

ndclt committed:
doc: fix linter

ndclt committed:
style: pep8

ndclt committed:
style: pylint issue

ndclt committed:
style: extra comma

ndclt commented:
I refactor a lot the code, I create a dedicated class for the keycloak user and create a new file `crud.py` managing crud for this class and future to come.

As suggested by @jamescassell, I replace the `credential` key by a `new_password` one.

ready_for_review

ndclt committed:
doc: improvement

tremble commented on lib/ansible/modules/identity/keycloak/keycloak_user.py:
```diff
@@ -0,0 +1,471 @@
+            - a dictionary with the key and the value to put in keycloak.
+              Keycloak will always return the value in a list of one element.
+              Keys and values are converted into string.
+        required: false
```
Required: false is implicit and not required

ndclt replied:
Changed in 525d5f2121c0558c972ffde3bae5fb3b2a29a9c1.

tremble commented on test/sanity/ignore.txt:
```diff
@@ -1822,6 +1822,7 @@ lib/ansible/modules/identity/keycloak/keycloak_clienttemplate.py validate-module
 lib/ansible/modules/identity/keycloak/keycloak_clienttemplate.py validate-modules:doc-required-mismatch
 lib/ansible/modules/identity/keycloak/keycloak_clienttemplate.py validate-modules:parameter-type-not-in-doc
 lib/ansible/modules/identity/keycloak/keycloak_group.py validate-modules:doc-required-mismatch
+lib/ansible/modules/identity/keycloak/keycloak_user.py validate-modules:doc-required-mismatch
```
It would be significantly better to fix the fragment / utility function : 

./lib/ansible/plugins/doc_fragments/keycloak.py line 25 lists auth_client_id as 'required'

keycloak_argument_spec (./lib/ansible/module_utils/identity/keycloak/keycloak.py) does not define it as required.

One of these is wrong and should be fixed rather than making things worse with yet more incorrect documentation.

ndclt replied:
Thank you for pointing that out. I the fragment. As there is a default value, it is not required. Changed in 40cb3046482de7e64f7068c5665cf8df5d690de4.

ndclt replied:
Solved with #66406.

ndclt committed:
doc: delete useless required: false

ndclt committed:
doc: fix not valid yaml

ndclt committed:
style: trailing whitespace...

ndclt commented:
bot_status

gundalow commented:
Hi
For this to be merged it would be great to hear some results of testing, ie have people tested this and does it work for you?

ndclt commented:
> Hi
> For this to be merged it would be great to hear some results of testing, ie have people tested this and does it work for you?

Hi,
People in my company use this module. Do you need comments from them approving this PR?

Nowheresly commented:
I wrote a similar PR (#67883) because I think ansible really lacks this feature. 
So why not testing this one :-)

I tested this PR by using the following docker-compose file:
```version: '3.3'

services:
  pgdb:
    image: postgres:12.2
    hostname: pgdb
    ports:
    - 5432:5432/tcp
    environment:
    - POSTGRES_PASSWORD=password
    volumes:
    - pgdb12-vol:/var/lib/postgresql/data

  keycloak:
    image: jboss/keycloak:9.0.0
    hostname: keycloak
    ports:
    - "8080:8080/tcp"
    environment:
      KEYCLOAK_USER: admin
      KEYCLOAK_PASSWORD: password
      DB_VENDOR: postgres
      DB_DATABASE: postgres
      DB_USER: postgres
      DB_PASSWORD: password
      DB_ADDR: pgdb
      DB_PORT: 5432

volumes:
  pgdb12-vol:
    driver: local
```

by running 
`docker-compose up -d`
we get a working instance of keycloak reachable at http://localhost:8080.
To log in, use admin/password.

Then, I  used the following playbook to create a new user:

```
---
- name: create user
  hosts: localhost
  gather_facts: false

  tasks:
  - name: Create a Keycloak user
    keycloak_user:
      keycloak_username: my-new-kc-user
      first_name: John
      last_name: Doe
      email: user-1@domain.com
      email_verified: True
      enabled: True
      realm: master
      state: present
      user_password: password
      required_actions:
        - UPDATE_PROFILE
      attributes:
        attrib1: value1
        attrib2: value2
      auth_client_id: admin-cli
      auth_keycloak_url: http://localhost:9191/auth
      auth_realm: master
      auth_username: admin
      auth_password: password
    delegate_to: localhost
```

Below are my comments:

- first of all, creation / update and delete of keycloak users are ok 👍 
- required_actions should allow `terms_and_conditions` and `update_user_locale` which are both valid keycloak required actions
- `attributes` field does not accept list as value like in module `keycloak_group`
Exemple (attrib3):
```
      attributes:
        attrib1: value1
        attrib2: value2
        attrib3:
          - with
          - numerous
          - individual
          - list
          - items
```
(yet, I'm not sure this is mandatory)
- finally, when creating a user, the username field is `keycloak_username` while the user password field is `user_password`. 
In my opinion, it would have been better to use the same convention for both fields. i.e either `keycloak_username` and `keycloak_password` or `user_username` and `user_password`.  As for me, the later makes more sense IMHO.

I hope this PR or mine (#67883 ) will be merged anytime soon!

ndclt commented:
@Nowheresly: thank you for testing this module. :-)
* What do you expect when giving a list as attributes of a user? All the values separated by comas?
* I will make arguments uniform with `user_username` and `user_password`.

Nowheresly commented:
>     * What do you expect when giving a list as attributes of a user? All the values separated by comas?

Just look at what have been done in module `keycloak_group`:

```
    if attributes is not None:
        for key, val in module.params['attributes'].items():
            module.params['attributes'][key] = [val] if not isinstance(val, list) else val

```

At the end, given a playbook like:
 ```
attributes:
        attrib1: value1
        attrib2: value2
        attrib3:
          - with
          - numerous
          - individual
          - list
          - items
```

Attributes in keycloak are:

- attrib1 = value1
- attrib2 = value2
- attrib3 = with##numerous##individual##list##items

ndclt committed:
fix: change name argument

ndclt committed:
feature: manage list as atttributes argument

ndclt committed:
typo: change the error message

ndclt committed:
typo: uncomment not done

ndclt committed:
doc: add element type for list

ndclt committed:
style: pep8

ndclt committed:
fix: put elements in required_actions of argument_sepc

Nowheresly commented:
Hi @ndclt I tested your PR with your last changes.
Same docker-compose as above and with the following playbook:

``` 
---
- name: create user
  hosts: localhost
  gather_facts: false

  tasks:
  - name: Create a Keycloak user
    keycloak_user:
      user_username: user2
      first_name: John
      last_name: Doe
      email: user-2@domain.com
      email_verified: True
      enabled: True
      realm: master
      state: present
      user_password: pass01
      required_actions:
        - UPDATE_PROFILE
#        - terms_and_conditions
#        - update_user_locale
      attributes:
        attrib1: value1
        attrib2: value2
        attrib3:
          - with
          - numerous
          - individual
          - list
          - items
      auth_client_id: admin-cli
      auth_keycloak_url: http://localhost:8080/auth
      auth_realm: master
      auth_username: admin
      auth_password: password
    delegate_to: localhost
``` 
Below are my updated comments:

- still no issue to create/update delete users :+1
- `attributes` field is now manage the same way as for `keycloak_group` 👍 
- both fields `user_username` and `user_password` have now a consistent naming :+1

but
- required_actions `terms_and_conditions` and `update_user_locale` are still missing 👎

mkrizek commented:
!waffling stale_ci

mkrizek commented:
Thank you for your interest in Ansible! We are no longer accepting new plugins to ansible/ansible directly. Please re-file this pull request against https://www.github.com/ansible-collections/community.general.

If you have further questions please stop by IRC or the mailing list:
* IRC: #ansible on irc.freenode.net
* mailing list: https://groups.google.com/forum/#!forum/ansible-project
